rfcs/text/0009_screenshot_mode_service.md

```diff
@@ -0,0 +1,260 @@
+- Start Date: 2020-03-02
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+The Kibana Reporting team has been chasing ways of improving performance of
+loading Kibana pages and capturing screenshots of the visualizations. We think
+that performance is a barrier to adding more integrations in Kibana, however we
+want to add more integrations and have Reporting be a factor of growth for
+Kibana. However, it looks like there are too many constraints to streamlining
+performance for Reporting if we only look for ways to change the Reporting code
+to do it. In other words, it will take a cross-team effort to improve the main
+performance issues. This RFC proposes one way to make cross-team efforts work
+smoothly, which is to have a new core service in Kibana that helps UIs and
+visualizations work nicely when loaded for screenshot capture.
+
+**Screenshot mode service**
+
+Applications in Kibana should be aware of when their rendering is for a human
+user using the page interactively, and when it's not. Sometimes the page is
+loaded for display-purposes: there is no user in control of the
+browser and no need for interactivity. Having Kibana support that purpose
+allows higher-level features to be built on top of rapidly-available images of
+Kibana pages. This turns out to be crucial for features like scheduled PDF
+report generation, and automated alerts with embedded images.
+
+As a bridge between the Kibana applications screenshot-capture features is
+needed as a core service in Kibana. This RFC names that service the Screenshot
```

Comment:
My primary critique of this RFC is about the assumption that this needs to be a Core service.

In general, we only consider including things in Core that either:
1) Must be present in order for Kibana to function at all (eg. config, HTTP server); or 
2) Is so widely used by most plugins that it is considered necessary (eg. UiSettings, SavedObjects); or
3) Is impractical to implement as a plugin or utility

I'm not sure that this fits into any of these criteria.

On the frontend, I believe this could be built as a simple utility that takes a `History` instance and returns an `Observable` that emits the screenshot data by listening for location updates and computing the data based on the query parameters.

```ts
const screenshot$ = plugins.screenshots.readFrom(history);
```

On the backend, this could be implemented as a wrapper around the `IRouter` interface that appends the `screenshot` query parameter to the `validation` option and then provides an extra property to `context` with the screenshot data. (This may be dependent on #59083).

```ts
const screenshotRouter = plugins.screenshots.wrapRouter(router);
screenshotRouter.get(
  { path: '/my-endpoint' },
  async (context, req, res) => {
    const isScreenshotMode = context.screenshots.isScreenshotMode();
  }
);
```

Benefits of implementing this as a plugin:
- If there is a bug in this service, not all of Kibana is broken
- This logic only needs to run for plugins that need it
- Keeps Core API small and lean
- Maintains the pattern on the client-side where we delegate all query parameters to the individual applications, letting them do as they wish with it.
- Less magic

Downsides:
- Plugins have to "wire-up" this service in their router and UI app
- May be less discoverable since this logic is hidden in a plugin rather than exposed on the `CoreSetup` API.

WDYT?

Reply:
I like the idea of this as a plugin, which lets us provide more than just a boolean in a URL param (IE auth helpers or other things if need be). Having to wire up something else in a plugin is a drawback, but I do think it opens up opportunities for us later if we can statically analyze which plugins are using reporting/screenshot-service. Might be able to reduce our/kbn's footprint versus making it a core service where it's always there.

@tsullivan thoughts on this? I'm leaning more plugin-oriented versus core-oriented on our architecture.

Reply:
I am ++ for the suggestion. I don't think there was a major reason for writing this up as a core service, just for the sake of brevity possibly.

Reply:
1711ecccfb